From 5bf1aba83ec980e23716144554d7547cbaf0a504 Mon Sep 17 00:00:00 2001
From: Alfonso Sanchez-Beato <alfonso.sanchez-beato@canonical.com>
Date: Mon, 21 Mar 2022 16:52:05 +0100
Subject: [PATCH] Do not enable for APs SAE if key management is psk

Enabling SAE when creating an access points make is work only when SAE
is supported by the connecting device, which is still not that
common. Disable then SAE if key management is psk: if wpa3 is desired
for the AP, "sae" needs to be set explicitly. Maybe this is a driver
issue and this patch will need to be reverted after consulting with
upstream.
---
 src/core/supplicant/nm-supplicant-config.c | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/src/core/supplicant/nm-supplicant-config.c b/src/core/supplicant/nm-supplicant-config.c
index 96c23579d..d3dd55227 100644
--- a/src/core/supplicant/nm-supplicant-config.c
+++ b/src/core/supplicant/nm-supplicant-config.c
@@ -35,6 +35,7 @@ typedef struct {
     bool           fast_required : 1;
     bool           dispose_has_run : 1;
     bool           ap_isolation : 1;
+    bool           is_ap : 1;
 } NMSupplicantConfigPrivate;
 
 struct _NMSupplicantConfig {
@@ -92,6 +93,7 @@ nm_supplicant_config_init(NMSupplicantConfig *self)
 
     priv->ap_scan         = 1;
     priv->dispose_has_run = FALSE;
+    priv->is_ap           = FALSE;
 }
 
 static gboolean
@@ -488,6 +490,7 @@ nm_supplicant_config_add_setting_wireless(NMSupplicantConfig *self,
     is_adhoc = nm_streq0(mode, "adhoc");
     is_ap    = nm_streq0(mode, "ap");
     is_mesh  = nm_streq0(mode, "mesh");
+    priv->is_ap = is_ap;
     if (is_adhoc || is_ap)
         priv->ap_scan = 2;
     else
@@ -854,10 +857,13 @@ nm_supplicant_config_add_setting_wireless_security(NMSupplicantConfig
             g_string_append(key_mgmt_conf, " WPA-PSK-SHA256");
         if (_get_capability(priv, NM_SUPPL_CAP_TYPE_FT))
             g_string_append(key_mgmt_conf, " FT-PSK");
-        if (_get_capability(priv, NM_SUPPL_CAP_TYPE_SAE)) {
-            g_string_append(key_mgmt_conf, " SAE");
-            if (_get_capability(priv, NM_SUPPL_CAP_TYPE_FT))
-                g_string_append(key_mgmt_conf, " FT-SAE");
+        /* If we want WPA3 for the AP, key_mgmt will need to be explicitly "sae" */
+        if (priv->is_ap == FALSE) {
+            if (_get_capability(priv, NM_SUPPL_CAP_TYPE_SAE)) {
+                g_string_append(key_mgmt_conf, " SAE");
+                if (_get_capability(priv, NM_SUPPL_CAP_TYPE_FT))
+                    g_string_append(key_mgmt_conf, " FT-SAE");
+            }
         }
 
     } else if (nm_streq(key_mgmt, "sae")) {
-- 
2.34.1

